Migrate upgrade tests to the new framework

mgencur opened the pull request:
Fixes #9137

<!-- Please include the 'why' behind your changes if no issue exists -->
## Proposed Changes

* Use the upgrade framework from knative.dev/pkg/test/upgrade
* Convert AssertAutoscaleUpToNumPods and inner functions to helper functions that return error instead of calling t.Fatal to fail the current test. Avoid using *testing.T so that these functions can be reused outside of tests or span multiple tests.
* The autoscaler helper functions are used in upgrade tests where "setup" and "verify" phases run within different tests. Pull test.EnsureTearDown from SetupSvc to ensure that a kservice is not destroyed at the end of the first phase ("setup") but remains active until "verify" phase. This is ensured by calling EnsureTearDown later in the  "verify" phase.
* Adjust Bash scripts to avoid unbound variable errors during upgrade tests exexution (due to using -u flag by the upgrade framework).

**Release Note**

<!-- Enter your extended release note in the below block. If the PR requires
additional action from users switching to the new release, include the string
"action required". If no release note is required, write "NONE". -->

```release-note

```

cardil commented on test/upgrade/autoscaler.go:
```diff
@@ -0,0 +1,96 @@
+		},
+		func(c pkgupgrade.Context) {
+			test.EnsureTearDown(c.T, ctx.Clients(), ctx.Names())
+			// Verification is done inside e2e.AssertAutoscaleUpToNumPods.
+			// We're just giving it a signal.
+			close(stopCh)
+			if err := grp.Wait(); err != nil {
+				c.T.Error("Error: ", err)
+			}
```
The whole point of this framework was to simplify those background verification tasks and their synchronization.

Maybe instead of doing here this "walkaround" a better idea would be to split `e2e.AssertAutoscaleUpToNumPods` to two parts - one doing setup, and one doing verification. Removing a stop channel, as there is already one in framework. That should simplify code a lot.

Also, for current code the function `pkgupgrade.NewBackgroundoperation` seams to be better suited, then simplified `pkgupgrade.NewBackgroundVerification`

mgencur replied:
The function `AssertAutoscaleUpToNumPods` is used in E2E tests in 8 other places. Splitting the function would mean more complicated usage in all the places and then maybe we'd save 2-3 lines in this upgrade test. I don't think it's worth it. Better to keep it straightforward in all the other test suites. Also, why do you think NewBackgroundOperation is more suited for this? I don't have any way to define what happens `OnWait` because that's buried in AssertAutoscaleUpToNumPods. I can't see the benefit of changing this. But maybe I'm missing something...

cardil replied:
Okay. I see.

Other proposal from me: Keep `e2e.AssertAutoscaleUpToNumPods` as it is, but rewrite it have two distinct parts separated with channel. Then use those parts (setup & verify) here, and on other places use `e2e.AssertAutoscaleUpToNumPods`.

Forget about using `NewBackgroundOperation` - you are also invoking, so that appropriate:
```golang
test.EnsureTearDown(c.T, ctx.Clients(), ctx.Names())
```

mgencur replied:
OK. I've pushed the split. It simplifies calling the e2e.AssertAutoscaleUpToNumPods in the other places. And the upgrade tests call the inner parts.

cardil replied:
Okay. Let's follow up that idea in separate follow up? Could you create an issue for that @mgencur?

mgencur replied:
Yeah. I'll create an issue for this.

mgencur replied:
https://github.com/knative/serving/issues/10276

cardil commented on test/upgrade/autoscaler.go:
```diff
@@ -0,0 +1,96 @@
+					autoscaling.TargetBurstCapacityKey: "0", // Not let Activator in the path.
+				}))
+			grp.Go(func() error {
+				return e2e.AssertAutoscaleUpToNumPods(ctx, logger.Sugar().Infof, 1, 10, stopCh, false /* quick */)
```
Aren't we like to control those parameters: `1, 10, false`?

mgencur replied:
They're constants. I'll just move them up to the `const` declaration.

cardil commented on test/upgrade/autoscaler.go:
```diff
@@ -0,0 +1,96 @@
+
+// AutoscaleSustainingWithTBCTest checks that when traffic increases and the activator is
+// in the path a knative app scales up and sustains the scale.
+func AutoscaleSustainingWithTBCTest(logger *zap.Logger) pkgupgrade.BackgroundOperation {
```
Here I have exactly the same comments as for `AutoscaleSustainingTest`

cardil commented on test/upgrade/probe.go:
```diff
@@ -0,0 +1,62 @@
+	v1test "knative.dev/serving/test/v1"
+)
+
+var successFraction = flag.Float64("probe.success_fraction", 1.0, "Fraction of probes required to pass during upgrade.")
```
Maybe this needs to be configured? My suggestion would be to put it in one configuration struct and pass it via `kelseyhightower/envconfig`, see: https://github.com/knative/eventing/blob/241af67/test/upgrade/prober/configuration.go#L94-L119

mgencur replied:
It's currently passed as a single flag. Why would we create a configuration struct for this? It sounds like you're proposing a different handling of parameters/arguments or environment and that should be tracked in a separate issue. This PR does not fix this.

cardil replied:
Fair enought

cardil commented on test/upgrade/upgrade_test.go:
```diff
@@ -0,0 +1,79 @@
+	suite := pkgupgrade.Suite{
+		Tests: pkgupgrade.Tests{
+			PreUpgrade: []pkgupgrade.Operation{
+				ServicePreUpgradeTest(),
+				ServicePreUpgradeAndScaleToZeroTest(),
+				BYORevisionPreUpgradeTest(),
+				InitialScalePreUpgradeTest(),
```
Maybe, create additional public methods for those operations. I see there are multiple ones for serving. Having single method like `ServingPreUpgradeTests() []pkgupgrade.Operation` would be much simpler to integrate on `knative/operator` and vendor level.

mgencur replied:
I don't understand. Do you mean wrap all these methods in a new single method that would return
```[]pkgupgrade.Operation{
				ServicePreUpgradeTest(),
				ServicePreUpgradeAndScaleToZeroTest(),
				BYORevisionPreUpgradeTest(),
				InitialScalePreUpgradeTest(),
			}
``` ?

cardil replied:
Yep. Exaclly.

cardil commented on test/upgrade/upgrade_test.go:
```diff
@@ -0,0 +1,79 @@
+		},
+		Installations: pkgupgrade.Installations{
+			Base: []pkgupgrade.Operation{
+				// Do nothing. The initial version is already installed by scripts
+				// together with additional test resources.
```
I think we should put here `installation.LatestRelease()` and make scripts do nothing instead. That would give us the ability to integrate that on `knative/operator` and vendor level.

mgencur replied:
The problem is with workarounds that exist for test resources. For example, right now we have disable_chaosduck bash function. This would need to be called separately from Golang -> Bash. And if we have other workaround in the future it will need a separate bash function, and then Golang wrappers. This seems to complicated instead of just using what was setup at the beginning.

mgencur replied:
But I'd like to know other opinions as well.

cardil replied:
Bot on `knative/operator` on on vendor we will not use serving e2e-upgrade-tests.sh script. Rather, we will have our own entrypoint script there. And then you will need copy & paste serving stuff to it, and manually keep it in sync.

If you moved that to a function then less code will be needed to maintain manually. Take a look how eventing entrypoint script is looking: https://github.com/knative/eventing/blob/6065b38/test/e2e-upgrade-tests.sh#L24-L43 It basically doesn't do anything.

mgencur replied:
Yeah. I know what you mean with respect to re-using this in knative/operator. I've looked at the test setup there. Currently, the installation and setup of test resources are quite different in knative/serving nad knative/operator: installing different images, labeling namespaces differently, waiting for different services/pods. I think the main issue is unifying the code at the level of Bash. I understand you want to enforce that by using these parts in upgrade tests here but this sounds like a problem that should be resolved separately (mainly within Bash). When the Bash functions are fixed/unified they can be re-used by the upgrade framework in knative/operator - when we get to implementing it there. 
However, I wouldn't complicate the Bash scripts and stuff on the knative/serving side right now. IMO, what is there now is the least intrusive change and is in line with the test setup for E2E, and doesn't need wrappers for workarounds.

cardil replied:
On the second thought we got `installation.LatestRelease()` and `installation.Head()` so we should be good on downstream.

cardil reviewed (changes requested):
Looks quite well. I found some minor issues.

mgencur commented on test/e2e/autoscale.go:
```diff
@@ -120,6 +122,7 @@ func getVegetaTarget(kubeClientset kubernetes.Interface, domain, endpointOverrid
 
 func generateTraffic(
 	ctx *TestContext,
+	logf logging.FormatLogger,
```
Actually, I had that as the first impl but then I realized the TestContext is something that should include things related to the Test. And the logger is something more general. So I pulled it out of it and passing it separately now:-/

mgencur replied:
I used the same type that is used in Serving by `prober`, that is:
```
// FormatLogger is a printf style function for logging in tests.
type FormatLogger func(template string, args ...interface{})
```

cardil replied:
I was proposing a `zap.Logger` because it is already available in `pkgupgrade.Context`, so we can reuse that. And it has more flexibility.

mgencur replied:
I pushed that change. Hopefully it's what you expected.

mgencur commented on test/e2e/autoscale.go:
```diff
@@ -61,7 +63,7 @@ const (
 type TestContext struct {
 	t                 *testing.T
 	clients           *test.Clients
-	names             test.ResourceNames
+	names             *test.ResourceNames
```
I noticed that other members of this struct are all passed by pointers. Aside from that, on the consumer side we mostly want the pointer instead of the real object. Then it's possible to call directly `test.EnsureTearDown(c.T, ctx.Clients(), ctx.Names())` instead of 
```
names := ctx.Names()
test.EnsureTearDown(c.T, ctx.Clients(), &names)
```

mgencur commented on test/upgrade/autoscaler.go:
```diff
@@ -0,0 +1,96 @@
+)
+
+const (
+	containerConcurrency = 6
+	targetUtilization    = 0.7
```
These are simply constants that are passed directly to `e2e.SetupSvc`. I don't see the need to wrap them in a specific config object.

mgencur commented:
/retest

mgencur commented:
/retest

mgencur commented:
/retest

mgencur commented:
/retest

mgencur commented:
The error in `pull-knative-serving-istio-stable-no-mesh-tls` doesn't seem to be related to this PR:
```
2020/12/02 11:33:53 Failed to setup DNS record: timed out waiting for the condition
```

mgencur commented:
/retest

mgencur commented:
I've reverted this PR to the last working state which includes moving the Logf stuff into TestContext but doesn't include further abstractions for the autoscaler tests; I was getting strange errors which I couldn't resolve, such as https://prow.knative.dev/view/gs/knative-prow/pr-logs/pull/knative_serving/10216/pull-knative-serving-istio-stable-no-mesh/1334566653832204288, right now I don't know how to resolve it. It's a cosmetic change so I'd suggest going forward with the current state.

cardil commented on test/e2e/autoscale.go:
```diff
@@ -262,7 +270,7 @@ func assertScaleDown(ctx *TestContext) {
 
 	// Account for the case where scaling up uses all available pods.
-	ctx.t.Log("Wait for all pods to terminate.")
+	ctx.t.Logf("Wait for all pods to terminate.")
```
`ctx.t.Logf` should have been a `ctx.logf` right?
```suggestion
	ctx.logf("Wait for all pods to terminate.")
```

cardil commented on test/e2e/autoscale.go:
```diff
@@ -281,24 +289,24 @@ func assertScaleDown(ctx *TestContext) {
 	}
 
-	ctx.t.Log("The Revision should remain ready after scaling to zero.")
+	ctx.t.Logf("The Revision should remain ready after scaling to zero.")
```
```suggestion
	ctx.logf("The Revision should remain ready after scaling to zero.")
```

cardil commented on test/e2e/autoscale.go:
```diff
@@ -281,24 +289,24 @@ func assertScaleDown(ctx *TestContext) {
 	}
 
-	ctx.t.Log("Scaled down.")
+	ctx.t.Logf("Scaled down.")
```
```suggestion
	ctx.logf("Scaled down.")
```

cardil reviewed (changes requested):
Couple on nits. BTW. Maybe a [README.md](https://github.com/knative/serving/blob/master/test/upgrade/README.md) needs to be updated a little?

cardil reviewed (approved):
/lgtm

mgencur commented:
/assign @nak3

nak3 commented on test/e2e/autoscale.go:
```diff
@@ -201,12 +209,11 @@ func SetupSvc(t *testing.T, class, metric string, target int, targetUtilization
 		Service: test.ObjectNameForTest(t),
 		Image:   autoscaleTestImageName,
 	}
-	test.EnsureTearDown(t, clients, &names)
```
Is this EnsureTearDown is no longer necessary? If so the comment line on this func:
> // It sets up EnsureTearDown to ensure that resources are cleaned up when the
// test terminates.

https://github.com/mgencur/serving-1/blob/migrate_upgrade_tests/test/e2e/autoscale.go#L205-L206
should be updated.

mgencur replied:
Right, I'll remove the comment.

mgencur replied:
Done.

nak3 commented on test/upgrade/autoscaler.go:
```diff
@@ -0,0 +1,94 @@
+	"knative.dev/serving/test"
+
+	pkgupgrade "knative.dev/pkg/test/upgrade"
+
```
The blank lines before and after `pkgupgrade "knative.dev/pkg/test/upgrade"` are not necessary.

nak3 reviewed (commented):
LGTM. Only a few minor comments from me.

nak3 commented:
/cc @chizhg @coryrc @chaodaiG 
Could you please help the review?

cardil commented:
/lgtm

mgencur commented:
Rebased the vendor/module.txt conflict. No other changes added.

markusthoemmes commented on test/e2e-common.sh:
```diff
@@ -15,8 +15,9 @@
-source $(dirname $0)/../vendor/knative.dev/hack/e2e-tests.sh
-source $(dirname $0)/e2e-networking-library.sh
+# shellcheck disable=SC1090
+source "$(dirname "${BASH_SOURCE[0]}")/../vendor/knative.dev/hack/e2e-tests.sh"
+source "$(dirname "${BASH_SOURCE[0]}")/e2e-networking-library.sh"
```
Let's please do changes like these separately. they unnecessarily bloat the diff and have nothing to do with the intended change as far as I can tell?

I believe this applies to most of the bash changes?

markusthoemmes replied:
If this is required for the shell shenanigans, let's still break the bash changes out to break this up a little.

mgencur replied:
This is required so that we can call e2e-common.sh from Golang. Without that the file cannot be found. I don't think this can be sent separately.

cardil replied:
This can be sent separately, but that has only little sense. As, apart from this use case previous code works, so without changes in here it's hard to justify changing that. Maybe say it's required for this PR, or it may error if sourced from different place than `test/` directory.

markusthoemmes commented on test/e2e/autoscale.go:
```diff
@@ -60,8 +62,9 @@ const (
 // TestContext includes context for autoscaler testing.
 type TestContext struct {
 	t                 *testing.T
+	logf              logging.FormatLogger
```
Break out this refactoring? Seems valid even at HEAD :thinking:

mgencur replied:
This is tightly coupled with the other changes in this PR. If I don't pass logf and use T.Infof directly the upgrade tests end with RACE errors because the Golang testing framework uses that concurrently with the function called from upgrade tests. I guess this could be sent in a separate commit in this PR but don't see a good reason for sending this in a separate PR.

markusthoemmes replied:
Right, but it can easily be frontloaded by making all the changes necessary on HEAD and by referencing the issue you're working on, stating exactly what you just said as the reason for the PR. It's even better than today as there' no explanation on why this is needed currently.

markusthoemmes commented on test/e2e/autoscale.go:
```diff
@@ -390,7 +404,5 @@ func AssertAutoscaleUpToNumPods(ctx *TestContext, curPods, targetPods float64, d
-	if err := grp.Wait(); err != nil {
-		ctx.t.Fatal("Error: ", err)
-	}
+	return grp
```
I think it's somewhat unusual to pass errgroups around. Should we return `func() error` instead and just return the `Wait` method of the group? Seems to have a similar effect.

mgencur replied:
OK. Why not.

markusthoemmes reviewed (commented):
This looks great :tada: !

Let's please invest in reviewability a bit though. There's quite a few refactorings here that I believe we can safely break out and send as their own PRs.

Top of mind:

1. The bash refactors as far as they make sense standalone
2. `ctx.logf` refactoring
3. `EnsureTearDown` additions

---

/hold

I'm applying a hold since we currently have a serious regression in upgrade tests at HEAD and I'd like to avoid having a situation where we don't know if they are due to a rewrite or not.

mgencur reviewed (commented):
> * The bash refactors as far as they make sense standalone

They're done for the purpose of calling e2e-common.sh from Golang. It makes sense to have it as part of this PR, IMO.

> * ctx.logf refactoring

See my explanation in a separate comment.

> * EnsureTearDown additions

This refactoring is required for running the autoscale tests by the upgrade test suite. It's an integral part of this PR.

mgencur commented:
/retest

markusthoemmes commented on test/e2e/autoscale.go:
```diff
@@ -367,15 +373,24 @@ func checkPodScale(ctx *TestContext, targetPods, minPods, maxPods float64, done
+	wait := AutoscaleUpToNumPods(ctx, curPods, targetPods, done, quick)
+	if err := wait(); err != nil {
+		ctx.t.Fatal(err)
+	}
+}
 
+// AutoscaleUpToNumPods starts the traffic for AssertAutoscaleUpToNumPods and returns
+// a function to wait for which will return any error from test execution.
+// Starting the routines is separated from waiting for easy re-use in other
+// places (e.g. upgrade tests).
+func AutoscaleUpToNumPods(ctx *TestContext, curPods, targetPods float64, done <-chan time.Time, quick bool) func() error {
 	// Relax the bounds to reduce the flakiness caused by sampling in the autoscaling algorithm.
 	// Also adjust the values by the target utilization values.
 	minPods := math.Floor(curPods/ctx.targetUtilization) - 1
 	maxPods := math.Ceil(math.Ceil(targetPods/ctx.targetUtilization) * 1.1)
 
 	stopChan := make(chan struct{})
-	var grp errgroup.Group
+	grp := &errgroup.Group{}
```
This change ain't required anymore.

markusthoemmes reviewed (commented):
As far as I can tell, this looks good to me! The tests failing intermittently while master was broken at least hints at the tests working as expected.

markusthoemmes commented:
Verified that the tests output looks like it's doing the correct things in the correct order too. Man, the output is awesome!

/lgtm

mgencur commented:
/test pull-knative-serving-istio-stable-no-mesh
/test pull-knative-serving-upgrade-tests

mgencur commented:
Added back the change to pass around `func() error` instead of `errGroup` as the behaviour is same and doesn't cause the latest test failures.
Added a flag for skipping Ingress upgrade. Previously the upgrade didn't happen either but it was hidden behind the NET_ISTIO_COMMIT flag. The variable was previously defined at the beginning and later reused so we were installing Ingress during the upgrade but it was the same version. The upgrade framework calls the install functions in separate shell so the variable is not reused during upgrade. That was causing the upgrade of Istio to a new version and tests were failing.

mattmoor commented on test/e2e/autoscale.go:
```diff
@@ -367,8 +373,17 @@ func checkPodScale(ctx *TestContext, targetPods, minPods, maxPods float64, done
 // from the given `done` channel will be sent within the `duration`.
 func AssertAutoscaleUpToNumPods(ctx *TestContext, curPods, targetPods float64, done <-chan time.Time, quick bool) {
-	ctx.t.Helper()
+	wait := AutoscaleUpToNumPods(ctx, curPods, targetPods, done, quick)
```
Why drop `t.Helper()`?

mgencur replied:
I don't mind dropping or keeping it. So added it back in a new commit, together with the other two occurrences.

mattmoor commented on test/e2e/autoscale.go:
```diff
@@ -367,8 +373,17 @@ func checkPodScale(ctx *TestContext, targetPods, minPods, maxPods float64, done
+// a function to wait for which will return any error from test execution.
+// Starting the routines is separated from waiting for easy re-use in other
+// places (e.g. upgrade tests).
+func AutoscaleUpToNumPods(ctx *TestContext, curPods, targetPods float64, done <-chan time.Time, quick bool) func() error {
```
should this have a `t.Helper()` too?

mattmoor commented on test/prober.go:
```diff
@@ -233,12 +233,18 @@ func RunRouteProber(logf logging.FormatLogger, clients *Clients, url *url.URL, o
 // against the default SLO, which requires perfect responses.
 // This takes `testing.T` so that it may be used in `defer`.
 func AssertProberDefault(t testing.TB, p Prober) {
+	AssertProberSLO(t, p, 1.0)
```
`t.Helper()`

mattmoor reviewed (commented):
Couple nits.  I'm not super familiar with the underlying library, but it seems @n3wscott reviewed it, and I'm game to try it out.  I like dropping some of the shell, but I don't love the Go delegating to shell stuff.  Do we think we'll replace more of that with Go over time?

thanks!

mgencur commented:
> Couple nits. I'm not super familiar with the underlying library, but it seems @n3wscott reviewed it, and I'm game to try it out. I like dropping some of the shell, but I don't love the Go delegating to shell stuff. Do we think we'll replace more of that with Go over time?
> 
> thanks!

Yes. We would like to migrate also the installation/upgrade from Bash to Golang over time. The Knative Eventing repo now delegates to Bash for these parts as well. The plan is to change it there as well. Step by step.

mgencur commented:
/test pull-knative-serving-istio-stable-no-mesh

mattmoor reviewed (commented):
/lgtm

nak3 commented:
/lgtm
/approve

mgencur commented:
/hold cancel

I believe the hold label from Markus is redundant at this point. The upgrade tests were fixed some time ago.

mgencur commented:
/retest
